File: python/a24.py
# A24 LIS
# submission: https://atcoder.jp/contests/tessoku-book/submissions/38565011

from bisect import bisect_left

n = int(input())
a = list(map(int, input().split()))

dp = [10 ** 6] * (n + 1) # dp[j]: 長さjのLISの最後の値として最小のもの
dp[0] = 0 # A[i] > 0 なので dp[0] = 0 とすると上手くいく
ans = 0

for i in range(n):
    j = bisect_left(dp, a[i]) # j: sortを崩さずa[i]を挿入できるindex <--> a[i]は長さjのLISの最後の値で最小のもの
    ans = max(ans, j)
    dp[j] = a[i]

print(ans)

File: rust/tessoku-book/src/bin/a24.rs
// A24 LIS
// submission: https://atcoder.jp/contests/tessoku-book/submissions/38565105

use proconio::input;
use std::cmp::max;

fn binary_search(x: &Vec<usize>, y: usize) -> usize {
    let mut l = 0;
    let mut r = x.len() - 1;
    while l < r {
        let m = (l + r) / 2;
        if x[m] < y {
            l = m + 1;
        }
        if x[m] >= y {
            r = m;
        }
    }
    return l;
}

fn main() {
    input!{
        n: usize,
        a:[usize; n],
    }
    let mut dp = vec![10_usize.pow(6); n + 1];
    dp[0] = 0;
    let mut ans = 0;
    for i in 0..n {
        let j = binary_search(&dp, a[i]);
        ans = max(ans, j);
        dp[j] = a[i];
    }
    println!("{}", ans);
}
